cocos/particle/line.ts

```diff
@@ -119,152 +117,147 @@ export class Line extends Component {
+            this._models.push(model);
+        } else {
+            this._models[0] = model;
+        }
         model.node = model.transform = this.node;
-        if (this._material === null) {
-            this._material = new Material();
-            this._material.copy(builtinResMgr.get<Material>('default-trail-material'));
+        if (this.lineMaterial === null) {
+            const mat = new Material();
+            mat.copy(builtinResMgr.get<Material>('default-trail-material'));
+            this.material = mat;
         }
-        if (this._material) {
+        const matIns = this.getMaterialInstance(0);
+        if (matIns) {
             define[CC_USE_WORLD_SPACE] = this.worldSpace;
-            _matInsInfo.parent = this._material;
-            _matInsInfo.subModelIdx = 0;
-            this._materialInstance = new MaterialInstance(_matInsInfo);
-            _matInsInfo.parent = null!;
-            _matInsInfo.subModelIdx = 0;
-            this._materialInstance.recompileShaders(define);
+            matIns.recompileShaders(define);
+            model.updateMaterial(matIns);
         }
-        model.updateMaterial(this._materialInstance!);
         model.setCapacity(100);
     }
 
     public onEnable () {
-        if (!this._model) {
+        super.onEnable();
+        if (this._models.length === 0 || !this._models[0]) {
             return;
         }
         this._attachToScene();
-        this.texture = this._texture;
-        this.tile = this._tile;
-        this.offset = this._offset;
-        this._model.addLineVertexData(this._positions, this._width, this._color);
+        const lineModel = this._models[0] as LineModel;
+        lineModel.addLineVertexData(this._positions, this.width, this.color);
     }
 
     public onDisable () {
-        if (this._model) {
+        if (this._models.length > 0 && this._models[0]) {
             this._detachFromScene();
         }
     }
 
     protected _attachToScene () {
-        if (this._model && this.node && this.node.scene) {
-            if (this._model.scene) {
+        super._attachToScene();
+        if (this._models.length > 0 && this._models[0] && this.node && this.node.scene) {
+            const lineModel = this._models[0];
+            if (lineModel.scene) {
                 this._detachFromScene();
             }
-            this._getRenderScene().addModel(this._model);
+            this._getRenderScene().addModel(lineModel);
         }
     }
 
     protected _detachFromScene () {
-        if (this._model && this._model.scene) {
-            this._model.scene.removeModel(this._model);
+        super._detachFromScene();
+        if (this._models.length > 0 && this._models[0]) {
+            const lineModel = this._models[0];
+            if (lineModel.scene) {
+                lineModel.scene.removeModel(lineModel);
+            }
+        }
+    }
+
+    protected update (dt: number): void {
+        if (this._models.length > 0 && this._models[0]) {
```

Comment:
Calcute the line vertices per frame is too wasted. Maybe update vertices when points changed.

Reply:
Changing value when use curve mode will not call the set function